pkg/agent/multicluster/pod_route_controller.go

```diff
@@ -0,0 +1,445 @@
+	podInfosMutex sync.Mutex
+	// ipToPods stores an IP and NamespacedName of all Pods which has been assigned with the same IP.
+	ipToPods      map[string]sets.String
+	ipToPodsMutex sync.Mutex
+	// started stores a boolean value, which tracks if the Pod flow controller has been started.
+	started  *atomic.Value
+	startCh  chan struct{}
+	finishCh chan struct{}
+}
+
+func NewMCPodRouteController(
+	k8sClient kubernetes.Interface,
+	mcClient mcclientset.Interface,
+	gwInformer cache.SharedIndexInformer,
+	client openflow.Client,
+	ovsBridgeClient ovsconfig.OVSBridgeClient,
+	interfaceStore interfacestore.InterfaceStore,
+	nodeConfig *config.NodeConfig,
+	namespace string,
+) *MCPodRouteController {
+	started := &atomic.Value{}
+	started.Store(false)
+	controller := &MCPodRouteController{
+		k8sClient:       k8sClient,
+		mcClient:        mcClient,
+		ovsBridgeClient: ovsBridgeClient,
+		ofClient:        client,
+		interfaceStore:  interfaceStore,
+		nodeConfig:      nodeConfig,
+		podQueue:        workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "MCPodRouteControllerForPod"),
+		gwQueue:         workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "MCPodRouteControllerForGateway"),
+		gwInformer:      gwInformer,
+		gwLister:        mclisters.NewGatewayLister(gwInformer.GetIndexer()),
+		namespace:       namespace,
+		podInfos:        make(map[types.NamespacedName]*podNodeIPs),
+		ipToPods:        make(map[string]sets.String),
+		started:         started,
+		startCh:         make(chan struct{}),
+		finishCh:        make(chan struct{}),
+	}
+
+	controller.gwInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(cur interface{}) {
+				controller.enqueueGateway(cur)
+			},
+			// Gateway UPDATE event doesn't impact Pod flows, so ignore it.
+			DeleteFunc: func(old interface{}) {
+				controller.enqueueGateway(old)
+			},
+		},
+		resyncPeriod,
+	)
+	return controller
+}
+
+func setPodInformer(controller *MCPodRouteController) {
+	listOptions := func(options *metav1.ListOptions) {
+		options.FieldSelector = fields.OneTermNotEqualSelector("spec.nodeName", controller.nodeConfig.Name).String()
+	}
+	controller.podInformer = coreinformers.NewFilteredPodInformer(
+		controller.k8sClient,
+		metav1.NamespaceAll,
+		0,
+		nil,
+		listOptions,
+	)
+	controller.podInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(cur interface{}) {
+				controller.enqueuePod(cur, false)
+			},
+			UpdateFunc: func(old, cur interface{}) {
+				controller.enqueuePod(cur, false)
+			},
+			DeleteFunc: func(old interface{}) {
+				controller.enqueuePod(old, true)
+			},
+		},
+		resyncPeriod,
+	)
+	controller.podLister = corelisters.NewPodLister(controller.podInformer.GetIndexer())
+}
+
+func handlePodFlowController(c *MCPodRouteController) {
```

Comment:
Question - why we need a Go routine to start/stop Pod controller, rather than do those directly in `syncGateway`?

Reply:
We need long running routine to make sure Pod flow controller will be running as long as there is Gateway change. If we put it in the syncGateway, the sub go-routine will be exit immediately when syncGateway is finished.
And we also need it for adding unit test easier.

Reply:
A goroutinue will not exist after its caller func exits. It exits only after the problem exits.

Reply:
Yeah, you are right regarding the go-routine part, I tried to move pod flow controller start process into `syncGateway`, there will be no way to control when the Pod event will be handled in the unit test because the `c.processPodNextWorkItem` in the Pod worker go-routine will run as long as the controller is started. There is no way to check expected flow installation because of timing issue. So I am afraid we still need this `handlePodFlowController` for unit test, With `handlePodFlowController`, we can skip this part and control it with a channel.

Reply:
I feel we should remove the `handlePodFlowController` goroutinue, which adds unnecessary complexity to the code. I do not understand your unit test problem, but you must have a way to implement UT without a goroutinue (e.g. I do not see people add goroutines for supporting UT). You can directly start/stop Pod controller when needed in your test code?

Reply:
`handlePodFlowController` can be removed if we move following part to `Run()` which means to keep a few go-routine for Pod always running without checking Gateway event. I will refine this part if you feel OK with this.
```
				for i := 0; i < defaultWorkerNum; i++ {
					go wait.Until(c.podWorker, time.Second, localStopCh)
				}
```

Reply:
I do not understand why that is the case. Could you explain why your UT code requires a long running goroutine? It does not make sense to me.

Reply:
No, I probably didn't clarify it clearly, UT doesn't need the go-routine part, The unit test skip this part by setting `started`, so this part will be skipped. I will move the go-rountine part  to `Run()`, then `handlePodFlowController` can be removed. I will refine this and update it. Thanks.

Reply:
To confirm - 1) we will not start watching Pods when the Node is not Gateway? 2) could you still explain why we must start the worker in `Run()` if that is what you mean. What is the real problem prevent you from starting the worker goroutines in `syncGateway`?

Reply:
1. Yes
2. if the worker goroutine is started in `syncGateway`, then following code will be run as background job for Gateway test cases:
```
	for c.processPodNextWorkItem() {
	}
```
Then we can't run test like below to check expected flow is installed or not because there is already a go-routine handle the event and timing for the result is unpredictable.
```
		c.ofClient.EXPECT().InstallMulticlusterPodFlows(nginx2PodIP, nginx2HostIP)
		c.processPodNextWorkItem()
```

Reply:
Please ignore my previous message, I think there was some errors when I tried yesterday which caused the test failure instead of go-routine issue. I will double check and update.

Reply:
This part is updated without `handlePodFlowController`, and remove the `startedMutex` since we will have only one go routine for Gateway event, and add a comment for `started` to make sure we will add a mutex if it's necessary in the future.